EIPS/eip-4337.md

```diff
@@ -159,6 +158,72 @@ interface IAggregator {
 * **validateSignatures()** MUST validate the aggregated signature matches for all UserOperations in the array, and revert otherwise.
   This method is called on-chain by `handleOps()`
 
+#### Semi-abstracted Nonce Support
+
+In Ethereum protocol, the sequential transaction `nonce` value is used as a replay protection method as well as to
+determine the valid order of transaction being included in blocks.
+
+It also contributes to the transaction hash uniqueness, as a transaction by the same sender with the same
+nonce may not be included in the chain twice.
+
+However, requiring a single sequential `nonce` value is limiting the senders' ability to define their custom logic
+with regard to transaction ordering and replay protection.
+
+Instead of sequential `nonce` we implement a nonce mechanism that uses a single `uint256` nonce value in the `UserOperation`,
+but treats it as two values:
+
+* 192-bit "key"
+* 64-bit "sequence"
+
+These values are represented on-chain in the `EntryPoint` contract.
+We define the following method in the `EntryPoint` interface to expose these values:
+
+```solidity
+function getNonce(address sender, uint192 key) external view returns (uint256 nonce);
```

Comment:
Can you clarify you design choice: If the intention was to pack the entire nonce = key || sequence, should the return value be whole nonce or the sequence?

Reply:
The return is the entire nonce (the current nonce for the user-provided key). It's what's going to be enforced during the operation itself.

Note that this section doesn't add anything new.  It was published as a separate document before the previous version, received good community feedback, and was subsequently added to EntryPoint.  The currently deployed version (v0.6) on mainnet already [implements](https://etherscan.io/address/0x5ff137d4b0fdcd49dca30c7cf57e578a026d2789#code#F4#L1) this.  We just forgot to update the EIP to include it 🤦